File: typescript/src/day01.ts
import { sumArray } from "./shared.ts";

const readCalories = () => {
  const input = Deno.readTextFileSync("../inputs/day01.txt");

  const calories = input.split("\n\n").map((elf) => {
    return elf.split("\n").reduce((sum, str) => {
      return sum + Number(str);
    }, 0);
  });

  return calories;
};

const calories = readCalories();

// sort descending
calories.sort((a, b) => b - a);

const max = calories[0];
console.log(`Part 1: ${max}`);

const top3 = sumArray(calories.slice(0, 3));
console.log(
  `Part 2: ${top3}`,
);

File: rust/src/bin/day01.rs
fn main() {
    let input = std::fs::read_to_string("../inputs/day01.txt").expect("Failed to read input file");

    let elves = input.split("\n\n");

    let mut calories = elves
        .map(|elf| {
            return elf
                .split("\n")
                .map(|line| line.parse::<u32>().expect("Failed to parse line"))
                .sum::<u32>();
        })
        .collect::<Vec<u32>>();

    let max = calories.iter().max().expect("Failed to find max");

    println!("Part 1: {:?}", max);

    calories.sort_by(|a, b| b.cmp(a));
    let top3 = calories.iter().take(3).sum::<u32>();

    println!("Part 2: {:?}", top3);
}
